### PR TITLE
Social links : Default background color problem fixed. Fix #34677

### LINKED ISSUE #34677: Social Icons: custom icon background colour does not persist when switching between block styles
### Description

If a background colour has been set for a social icon using the Default custom style, it does not persist when switching between the block's custom styles: Default and Logos Only.

I was surprised to see the the Default background colour was not reinstated when switching between styles but am unsure if this is intentional behaviour.

### Step-by-step reproduction instructions

1. Insert a Social Icons block
2. Add some icons.
3. Switch between the Styles presets — from Default to Logos Only, and back to Default  — and notice the default icon colours are retained.
4.  Set a custom icon colour and background colour via the _Icon background color_ picker.
5. Repeat step 3. Notice the background colour is not retained when returning to the Default style.

### Screenshots, screen recording, code snippet

Default icon colours:

social-icons-switch-styles-default-colours

Custom icon colours:

social-icons-switch-styles-custom-colours

### Environment info

- WP 5.8, Gutenberg 11.4.1, theme in screen caps is custom but can replicate in TT1 Blocks.
- Chrome 93.0.4577.63
- Windows 10 on desktop

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by HILAYTRIVEDI**
I tested the code with different themes which are default provided by WordPress. No other functionalities are affected by the changes. Two new `useEffect()` is added to the `packages\block-library\src\social-links/edit.js` for the default and pill classes. As in the pill also the same issue arises. 

After the changes, I tested it with the different combinations of the use of the styling pattern, and it's working fine.

## Video Demo Of Fix
https://www.awesomescreenshot.com/video/5232616?key=888676e6a53d9db85397b4430d64c810

## Types of changes

The use effect in the file uses the previous state of the `iconBackgroundColorValue` and while the class changes it uses the value of `preveIconBackgroundColorValue` for the previously set value of the background. 

Bug fix #34677

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js]: I suggest keeping this in state instead of block attributes. Also, because they are also set to `undefined` by this effect hook, it seems appropriate to also put `iconBackgroundColor` and `customIconBackgroundColor` values into the same state in order to also restore those when the `logosOnly` is no longer true.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js:90]:
```suggestion
		} else {
			// Restore previous background attributes.
			setAttributes( { ...backgroundBackup } );
		}
```
Instead of adding two new effect hooks that restore previous attributes it should work to put in an else clause here to do it.

**Review by stokesman — COMMENTED** [review]
Hi @HILAYTRIVEDI, thanks taking this on. It looks like this resolves the issue. I think it'd better done without adding another block attribute because those are meant for things that need to persist across saves. Besides that, I think we can make this a bit simpler. I've left a couple of comments to elaborate but let me know if anything isn't clear.

File: packages/block-library/src/social-links/edit.js
Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js:90]: Thank you for the feedback @stokesman. This is my first contribution and I appreciate your guidance. I did the changes in the latest PR.

File: packages/block-library/src/social-links/edit.js
Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js]: By putting it in to the state it will reset on refresh. So its better to create new attribute instead.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js]:
```suggestion
	const backgroundBackup = useRef( {} );
```
It's not crucial but I've realized `useRef` may be better here since another render is not needed when the value changes. Besides that, I'd put this on a line that's closer to the hook it's used in as that seems more consistent with the style of the rest of the codebase.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js]:
```suggestion
			backgroundBackup.current = {
				iconBackgroundColor,
				iconBackgroundColorValue,
				customIconBackgroundColor,
			};
```
To continue from the comment about switching to a ref.

File: packages/block-library/src/social-links/block.json
Comment (stokesman) [inline packages/block-library/src/social-links/block.json]:
```suggestion
		"customIconColor": {
```
Don't know why the linter doesn't complain about this but let's get rid of the tab.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js]:
```suggestion
	// Remove icon background color when logos only style is selected or
	// restore it when any other style is selected.
```
Just to mention the added behavior in the comment. I also think this comment would go better on the lines just before the the effect hook but I suppose it's okay where we found it.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js:91]: As is, the effect could be considered to be missing dependencies (`iconBackgroundColor` and the other two that are put into backup). But I've tested and I can't make it misbehave (i.e. backup stale values). If we add the dependencies it would require more conditional logic to avoid breaking the effect. There are other options that avoid referencing those variables inside the hook but I'm not sure any of that complexity is worth it when this seems okay for now.

**Review by stokesman — COMMENTED** [review]
Thanks for the updates @HILAYTRIVEDI 👍  It's much nicer with the reduced changes. I've left some more feedback. 

It'd be great to have another reviewer have a look but I see some others have been requested so I'll give them some time to get to it before pinging them again.

File: packages/block-library/src/social-links/edit.js
Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js:91]: @stokesman thanks for the feedback but I think its okay of now.

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [inline packages/block-library/src/social-links/edit.js:91]: So this doesn't seem necessary (as mentioned before: https://github.com/WordPress/gutenberg/pull/34834#discussion_r715202668) but it may be better to have all the consumed variables in the hook dependencies and it doesn't end up adding much complexity.
```suggestion
			setAttributes( {
				iconBackgroundColor: undefined,
				customIconBackgroundColor: undefined,
				iconBackgroundColorValue: undefined,
			} );
			// Restore the attributes from backup on cleanup.
			return () => setAttributes( { ...backgroundBackup.current } );
		}
		// Backup icon background attributes.
		backgroundBackup.current = {
			iconBackgroundColor,
			iconBackgroundColorValue,
			customIconBackgroundColor,
		};
	}, [
		logosOnly,
		setAttributes,
		iconBackgroundColor,
		iconBackgroundColorValue,
		customIconBackgroundColor,
	] );
```

**Review by stokesman — APPROVED** [review]
This works well in my testing and I've left one more suggestion but it seems optional. Perhaps another reviewer can weigh in on that. Either way, I think we could merge this. Thanks @hilaytrivedi!

File: packages/block-library/src/social-links/edit.js
Comment (stokesman) [reply packages/block-library/src/social-links/edit.js:91]: Yes, I was saying as much. Still, I took another look and found a pretty simple way to have it work with the added dependencies. I've left that as separate comment/suggestion.

Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js:91]: Thank you soo much @stokesman for the feedbacks as I am new at contribution and I learned some new things by your comments.

File: packages/block-library/src/social-links/edit.js
Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js:91]: Agreed.

Comment (HILAYTRIVEDI) [reply packages/block-library/src/social-links/edit.js:91]: @stokesman I am happy that I get chance for contributing.

**Comment by Mamaduka** [comment]
Hi, @HILAYTRIVEDI

Can you [rebase](https://developer.wordpress.org/block-editor/contributors/code/git-workflow/#keeping-your-branch-up-to-date) this branch on top of the trunk? I think we can merge this once all checks are green.

**Comment by HILAYTRIVEDI** [comment]
Hi @Mamaduka , I am unable to rebase the branch.

**Pull request closed without merging**
